server/src/main/java/org/elasticsearch/cluster/ClusterModule.java

```diff
@@ -251,4 +252,21 @@ protected void configure() {
         bind(AllocationDeciders.class).toInstance(allocationDeciders);
         bind(ShardsAllocator.class).toInstance(shardsAllocator);
     }
+
+    public void setExistingShardsAllocators(GatewayAllocator gatewayAllocator, List<ClusterPlugin> clusterPlugins) {
```

Comment:
We have to defer the creation of the allocators until later, because they need a `RerouteService` and there's an (intrinsically tricky) circular dependency to resolve. We'll need to do this differently when the plugin's allocator also needs a `RerouteService` but it's premature to address that  now.

Reply:
Ok, addressed in 4b8888c0311d0bfc886baa7ae0349b61749d103e.